frame/contracts/src/schedule.rs

```diff
@@ -53,20 +57,20 @@ pub struct Schedule<T: Config> {
 	/// of all contracts which are triggered by a version comparison on call.
 	/// Changes to other parts of the schedule should not increment the version in
 	/// order to avoid unnecessary re-instrumentations.
-	pub version: u32,
+	pub(crate) version: u32,
```

Comment:
I'm not it makes complete sense to me to make those fields private.
The type is part of the API of contract calls, I would expect it to be completly public.
At some point I would expect any type which are part of the pallet API (calls, events, origins) to have their codec encoding part of the public API of the crate

It is also part of the genesis config, don't we want to allow user to be able to use other value than default?

Reply:
Making those fields public will require us to release a new major version every time we add a new contract callable function. Each of those require a new field in the `Schedule`. This is why I made them private. For the genesis part: We can add more constructors which allow more tuning to the fields. As an extreme we could have constructors which just take an argument for every field. When we update the `Schedule` with a new field we add a new constructor and the old one just initializes the new field with its default value. But I leave that as future work until we learn how users want to use this.

> The type is part of the API of contract calls, I would expect it to be completly public.

Why? Types with private fields but public constructors and getters are not uncommon. The `set_schedule` extrinsic will be crafted by javascript code anyways I assume. Having the fields private to Rust allows us to be more flexible with versioning.

Reply:
> > The type is part of the API of contract calls, I would expect it to be completly public.
>
> Why? Types with private fields but public constructors and getters are not uncommon. The set_schedule extrinsic will be crafted by javascript code anyways I assume. Having the fields private to Rust allows us to be more flexible with versioning.

yes ok, but I feel like now people who are crafting extrinsics have no documentation about what is the type `Schedule`.
but I agree the documentation should be elsewhere.

Also sidenote but for me it would make sense to consider the encoding of `Call` as part of the public API.
Any breakage in the encoding of `Call` requires to bump the transaction version of the runtime and any tools which uses the call encoding.

Reply:
I disagree. Semver should be about making your `cargo update` experience as hassle free as possible. Bumping a major version all the time because you make the public API definition to bread is detrimental to that goal.

For the off-chain environment (e.g tools) the semver is not helpful anyways. They need to be updated when the runtime version changes. They don't use semver for this but release notes. Those should mention all the information they need to update their tools (like transaction version or storage breakage). The release notes are generated from github labels and not semver.

> Any breakage in the encoding of Call requires to bump the transaction version of the runtime and update any tools which uses the call encoding.

Does this mean changing the `Schedule` needs a transaction version bump (because it is used in a `Call`)?

Reply:
> Does this mean changing the Schedule needs a transaction version bump (because it is used in a Call)?

Yes AFAIU, changing the encoding of schedule change the definition of runtime transactions.

> Semver should be about making your `cargo update` experience as hassle free as possible

It makes sense to me to consider `<Call as Decode>::decode` part of the public API, and changing it does change the API in a not backward compatible manner.

Reply:
> It makes sense to me to consider <Call as Decode>::decode part of the public API, and changing it does change the API in a not backward compatible manner.

If this is true making the fields of the `Schedule` does not change anything with regard to semver. I really don't want to burn through major versions that quickly. I am reluctant to make the javascript side part of the public API of the Rust program. IMHO semver should only cover the Rust side. The decode part should be covered by transaction version as you mentioned.

Reply:
I agree to have the field of schedule private, I hope we can find a way to document those stuff for other developer.

I still think `<Call as Decode>::decode` is part of rust side, if I change the implementation of decode for `u8` in parity-scale-codec IMO we need to bump major version of parity-scale-codec. I don't see why the situation is fundamentally different here.

But if it is only me then it is fine, I don't mind. But then runtime will have to be careful that updating a minor version of a crate might require to change the `spec_version` and `transaction_version` of the runtime.

Reply:
In a follow up PR I will remove the `Schedule` from storage and simply make it part of the `Config` trait. This will resolve this issue. I cannot really accomplish it in this PR because that is a bit more involved.